internal/daemon/worker/worker.go

```diff
@@ -423,6 +423,38 @@ func (w *Worker) Start() error {
 	return nil
 }
 
+func (w *Worker) hasActiveConnection() bool {
+	activeConnection := false
+	w.sessionManager.ForEachLocalSession(
+		func(s session.Session) bool {
+			conns := s.GetLocalConnections()
+			for _, v := range conns {
+				if v.Status == pbs.CONNECTIONSTATUS_CONNECTIONSTATUS_CONNECTED {
+					activeConnection = true
+					return false
+				}
+			}
+			return true
+		})
+	return activeConnection
+}
+
+// Graceful shutdown sets the worker state to "shutdown" and will wait to return until there
+// are no longer any active connections.
+func (w *Worker) GracefulShutdown() error {
+	const op = "worker.(Worker).GracefulShutdown"
+	event.WriteSysEvent(w.baseContext, op, "worker entering graceful shutdown")
+	w.operationalState.Store(server.ShutdownOperationalState)
+
+	// Wait for connections to drain
+	for w.hasActiveConnection() {
+		time.Sleep(time.Second)
```

Comment:
Assuming that GetLocalConnections doesn't make a database call, I think you could actually do this more often, maybe 250ms. But we can always tweak later!